src/api.c

```diff
@@ -536,7 +548,20 @@ static struct nmd_frame *ret_frame(struct nmd_ctx *s, AVFrame *avframe, int stat
 
 end:
     END_FUNC(MAX_SYNC_OP_TIME);
-    return frame;
+
+    *framep = frame;
+
+    int ret;
+    if (frame)
+        ret = NMD_RET_NEWFRAME;
+    else if (status >= 0)
+        ret = NMD_RET_UNCHANGED;
+    else if (s->eof)
+        ret = NMD_ERR_EOF;
+    else
+        ret = averror_to_nmd_err(status);
+
+    return ret;
```

Comment:
nit: you could return directly in each branch (doesn't even require all the `else`) and avoid that late `int ret`

Reply:
Changed.